### PR TITLE
tracing: Add hostname to Zipkin config.  (#14186)

### LINKED ISSUE #14186: Support setting hostname on the Zipkin collector config
Envoy should support setting the hostname used when sending spans to a Zipkin collector. Currently, the cluster name is used, but this isn't always the hostname that the upstream collector wants.

This should be a straightforward change to the Zipkin tracing config and the `zipkin_tracer_impl`.

### PR DESCRIPTION
**Pull request by esmet**
This should be ready for review, though docs and release notes are still needed. Feedback on all fronts welcome.

Commit Message:  tracing: Add hostname to Zipkin config
Additional Description:
Risk Level: low (new configuration, default behavior preserved)
Testing: unit (expanded an existing test)
Docs Changes: API protos documented, in line with existing fields
Release Notes: tracing: added support for setting the hostname used when sending spans to a Zipkin collector using the :ref:`collector_hostname <envoy_v3_api_field_config.trace.v3.ZipkinConfig.collector_hostname>` field.
Fixes #14186

### CODE REVIEW

**Comment by esmet** [comment]
I force pushed an update to correct a DCO mistake per the guidelines.

**Comment by mk46** [comment]
Need to fix the format. Run `tools/code_format/check_format.py fix` to fix it

**Comment by esmet** [comment]
> tools/code_format/check_format.py fix

good call, done

**Review by dio — COMMENTED** [review]
Looks good. Please provide release notes and docs. Thanks!

**Comment by esmet** [comment]
@dio the release notes were easy enough 👍 

As for docs, it looks like we have a sample Envoy config `./examples/zipkin-tracing/front-envoy-zipkin.yaml` but no additional documentation around the config itself. My understanding is that we typically rely on the API proto comments/docs for that. Is that right? And if not, what would you recommend?

File: api/envoy/config/trace/v3/zipkin.proto [hunk 1]
```diff
@@ -65,4 +65,8 @@ message ZipkinConfig {
   // Determines the selected collector endpoint version. By default, the ``HTTP_JSON_V1`` will be
   // used.
   CollectorEndpointVersion collector_endpoint_version = 5;
+
+  // Optional hostname to use when sending spans to the collector_cluster. Useful for collectors
+  // that require a specific hostname. Defaults to `collector_cluster` above.
```
Comment (dio) [inline api/envoy/config/trace/v3/zipkin.proto:70]: We usually link this to the actual field.

Comment (esmet) [reply api/envoy/config/trace/v3/zipkin.proto]: Done. I did something similar to the link I added in the release notes, let me know if that's what you had in mind.

**Comment by dio** [comment]
Can you check CI? Seems like "verify examples" failed on Zipkin.

**Comment by esmet** [comment]
> Can you check CI? Seems like "verify examples" failed on Zipkin.

I looked at this but can't figure out why it failed. I see some configuration errors though:
```
2020-12-02T22:25:27.9159539Z [36mvrp_1  |[0m (edge)[2020-12-02 22:25:17.813][11][critical][main] error initializing configuration '/etc/envoy/envoy-edge.yaml': Unable to parse JSON as proto (INVALID_ARGUMENT:(static_resources.clusters.typed_extension_protocol_options[0].value): invalid value Invalid type URL, unknown type: envoy.extensions.upstreams.http.v3.HttpProtocolOptions for type Any): {"static_resources":{"clusters":{"typed_extension_protocol_options":{"envoy.extensions.upstreams.http.v3.HttpProtocolOptions":{"explicit_http_config":{"http2_protocol_options":{"initial_connection_window_size":1048576,"initial_stream_window_size":65536}},"@type":"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions"}},"connect_timeout":"5s","load_assignment":{"cluster_name":"service_foo","endpoints":[{"lb_endpoints":[{"endpoint":{"address":{"socket_address":{"port_value":10002,"address":"127.0.0.1"}}}}]}]},"per_connection_buffer_limit_bytes":32768,"name":"service_foo"},"listeners":[{"filter_chains":[{"filters":[{"typed_config":{"stream_idle_timeout":"300s","route_config":{"name":"local_route","virtual_hosts":[{"routes":[{"route":{"idle_timeout":"15s","cluster":"service_foo"},"match":{"prefix":"/content"}},{"direct_response":{"body":{"inline_string":"denied\n"},"status":403},"match":{"prefix":"/"}}],"name":"local_service","domains":["*"]}]},"use_remote_address":true,"common_http_protocol_options":{"idle_timeout":"3600s","headers_with_underscores_action":"REJECT_REQUEST"},"@type":"type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager","http2_protocol_options":{"max_concurrent_streams":100,"initial_connection_window_size":1048576,"initial_stream_window_size":65536},"request_timeout":"300s","http_filters":[{"name":"envoy.filters.http.router"}],"stat_prefix":"ingress_http"},"name":"envoy.filters.network.http_connection_manager"}],"transport_socket":{"typed_config":{"@type":"type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext","common_tls_context":{"tls_certificates":[{"private_key":{"filename":"certs/serverkey.pem"},"certificate_chain":{"filename":"certs/servercert.pem"}}]}},"name":"envoy.transport_sockets.tls"}}],"name":"listener_https","per_connection_buffer_limit_bytes":32768,"address":{"socket_address":{"address":"0.0.0.0","protocol":"TCP","port_value":10000}}}]},"overload_manager":{"resource_monitors":[{"name":"envoy.resource_monitors.fixed_heap","typed_config":{"@type":"type.googleapis.com/envoy.config.resource_monitor.fixed_heap.v2alpha.FixedHeapConfig","max_heap_size_bytes":1073741824}}],"refresh_interval":"0.25s","actions":[{"triggers":[{"name":"envoy.resource_monitors.fixed_heap","threshold":{"value":"0.90"}}],"name":"envoy.overload_actions.shrink_heap"},{"name":"envoy.overload_actions.stop_accepting_requests","triggers":[{"threshold":{"value":"0.95"},"name":"envoy.resource_monitors.fixed_heap"}]}]}}
...
```

I'll try rerunning to see if this reproduces.

**Comment by esmet** [comment]
> > Can you check CI? Seems like "verify examples" failed on Zipkin.
> 
> I looked at this but can't figure out why it failed. I see some configuration errors though:
> 
> ```
> 2020-12-02T22:25:27.9159539Z �[36mvrp_1  |�[0m (edge)[2020-12-02 22:25:17.813][11][critical][main] error initializing configuration '/etc/envoy/envoy-edge.yaml': Unable to parse JSON as proto (INVALID_ARGUMENT:(static_resources.clusters.typed_extension_protocol_options[0].value): invalid value Invalid type URL, unknown type: envoy.extensions.upstreams.http.v3.HttpProtocolOptions for type Any): {"static_resources":{"clusters":{"typed_extension_protocol_options":{"envoy.extensions.upstreams.http.v3.HttpProtocolOptions":{"explicit_http_config":{"http2_protocol_options":{"initial_connection_window_size":1048576,"initial_stream_window_size":65536}},"@type":"type.googleapis.com/envoy.extensions.upstreams.http.v3.HttpProtocolOptions"}},"connect_timeout":"5s","load_assignment":{"cluster_name":"service_foo","endpoints":[{"lb_endpoints":[{"endpoint":{"address":{"socket_address":{"port_value":10002,"address":"127.0.0.1"}}}}]}]},"per_connection_buffer_limit_bytes":32768,"name":"service_foo"},"listeners":[{"filter_chains":[{"filters":[{"typed_config":{"stream_idle_timeout":"300s","route_config":{"name":"local_route","virtual_hosts":[{"routes":[{"route":{"idle_timeout":"15s","cluster":"service_foo"},"match":{"prefix":"/content"}},{"direct_response":{"body":{"inline_string":"denied\n"},"status":403},"match":{"prefix":"/"}}],"name":"local_service","domains":["*"]}]},"use_remote_address":true,"common_http_protocol_options":{"idle_timeout":"3600s","headers_with_underscores_action":"REJECT_REQUEST"},"@type":"type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager","http2_protocol_options":{"max_concurrent_streams":100,"initial_connection_window_size":1048576,"initial_stream_window_size":65536},"request_timeout":"300s","http_filters":[{"name":"envoy.filters.http.router"}],"stat_prefix":"ingress_http"},"name":"envoy.filters.network.http_connection_manager"}],"transport_socket":{"typed_config":{"@type":"type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.DownstreamTlsContext","common_tls_context":{"tls_certificates":[{"private_key":{"filename":"certs/serverkey.pem"},"certificate_chain":{"filename":"certs/servercert.pem"}}]}},"name":"envoy.transport_sockets.tls"}}],"name":"listener_https","per_connection_buffer_limit_bytes":32768,"address":{"socket_address":{"address":"0.0.0.0","protocol":"TCP","port_value":10000}}}]},"overload_manager":{"resource_monitors":[{"name":"envoy.resource_monitors.fixed_heap","typed_config":{"@type":"type.googleapis.com/envoy.config.resource_monitor.fixed_heap.v2alpha.FixedHeapConfig","max_heap_size_bytes":1073741824}}],"refresh_interval":"0.25s","actions":[{"triggers":[{"name":"envoy.resource_monitors.fixed_heap","threshold":{"value":"0.90"}}],"name":"envoy.overload_actions.shrink_heap"},{"name":"envoy.overload_actions.stop_accepting_requests","triggers":[{"threshold":{"value":"0.95"},"name":"envoy.resource_monitors.fixed_heap"}]}]}}
> ...
> ```
> 
> I'll try rerunning to see if this reproduces.

I think this could be elated to a merge to master that happened yesterday. I'll rebase. If CI goes green, @dio do you think the release notes change and API comments are enough to land this?

**Comment by esmet** [comment]
I rebased and CI looks good now. This should be ready for final review :)

**Review by dio — APPROVED** [review]
Thanks. Change looks good. However, this needs @envoyproxy/api-shepherds approval for API.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Pull request merged**